New lesson on using flask to create apis

walshbr opened the pull request:
Note: not ready for review and merging just yet. Using the netlify preview for soft launch.

walshbr committed:
Add initial files

walshbr committed:
Add Smyth author bio

walshbr committed:
Add yaml

walshbr committed:
Upload images

walshbr committed:
Delete duplicate toc since we use an include

walshbr committed:
Update author metadata for smyth

walshbr committed:
Update author and abstract metadata

walshbr committed:
Resolve type

walshbr commented:
This should be ready to merge and publicize.

mdlincoln commented:
I'll take a look!

https://deploy-preview-773--ph-dev.netlify.com/lessons/creating-apis-with-python-and-flask

ianmilligan1 commented:
I'll take a quick review too 👍

ianmilligan1 commented on lessons/creating-apis-with-python-and-flask.md:
```diff
@@ -0,0 +1,713 @@
+
+To download Python, follow [this link](https://www.python.org/downloads/), select the button that says `Download Python 3.x.x`, and then run the installer as you normally would to install applications on your operating system. The default settings should be fine.
+
+To confirm that Python installed successfully, first open the command line. In OSX, click the finder on the top left of your desktop (the magnifying glass) and type `terminal`. The terminal should be the first application that appears. On Windows, click the Start menu icon and type `cmd` in the search box, then press `Enter`.
```
Pedantic: "finder on the top left of your desktop (the magnifying glass)" --> this is "spotlight," right? Finder is the file browser, and spotlight is in the upper-right corner. 😄 

Maybe change to

"In macOS, click the spotlight icon on the top right corner of your desktop (the magnifying glass) and type `terminal`."

walshbr replied:
Resolved!

ianmilligan1 commented on lessons/creating-apis-with-python-and-flask.md:
```diff
@@ -0,0 +1,713 @@
+
+## Prerequisites
+
+You can use the Windows, OSX, or Linux operating systems to complete this tutorial, and those few instructions that are not the same across platforms will be explicitly noted. Python 3, the Flask web framework, and a web browser are required for this tutorial, and installation instructions for all platforms are outlined below.
```
OK I only have pedantic little things.

OS X has been frustratingly renamed to macOS. Might be worth doing a find and replace.

walshbr replied:
Resolved!

ianmilligan1 commented on lessons/creating-apis-with-python-and-flask.md:
```diff
@@ -0,0 +1,713 @@
+
+The above phone book is a list of two dictionaries. Each dictionary is a phone book entry consisting of two keys, `name` and `number`, each paired with a value that provides the actual information.
+
+Let's add some data—entries on three science fiction novels—as a list of dictionaries. Each dictionary will contain ID number, title, author, first sentence, and year of publication for each book. Finally, we'll add a new function: a route that will allow a visitor to access our data.
```
data-entries -> data entries

walshbr replied:
Resolved!

ianmilligan1 commented on lessons/creating-apis-with-python-and-flask.md:
```diff
@@ -0,0 +1,713 @@
+
+## What Users Want in an API
+
+As we've learned, documentation is a user's starting place when working with a new API, and well-designed URLs make it easier for users to intuitively find resources. Because they help users to quickly access information through your API, these elements—documentation and well-conceived URLs—are the *sine qua non* of a good API. We'll discuss these elements in greater depth later in this tutorial.
```
*sine qua non* - I'm a ignoramus who sometimes stumbles over things like this. Could we just say "The essential element"?

walshbr replied:
Deferring to Patrick.

mdlincoln committed:
use relative links to other PH pages

ianmilligan1 commented on lessons/creating-apis-with-python-and-flask.md:
```diff
@@ -0,0 +1,713 @@
+
+As new Hugo winners were added to the database, the script that generated this visualization would immediately be able to use the new information. If the visualization were created in D3 or another web-based utility, this plot would actually reflect additional data added to the book archive as soon as the archive was updated—that is, in real time. As additional data accrued, we might, for example ...
+
+A strong API can be considered the backbone of a potentially limitless number of projects or avenues of research. Though the above example takes the form of a visualization of the limited amount of data we've provided in our Distant Reading Archive, a project based on this API might just as easily take the form of a Twitterbot that shares first sentences or a library webpage that displays book op ...
```
For Twitterbot, make sense to link to this _Programming Historian_ lesson?

https://programminghistorian.org/lessons/intro-to-twitterbots

ianmilligan1 reviewed (approved):
This is a fantastic lesson, and as somebody who's been hoping to learn flask I can't wait to give this a try soon.

I left a few comments here, none of which are dealbreakers - but the author could consider them before it goes live.

mdlincoln commented on lessons/creating-apis-with-python-and-flask.md:
```diff
@@ -0,0 +1,713 @@
+
+[History Data Service (HDS)](http://hds.essex.ac.uk/) - A collection of data from a wide variety of historical sources.
+
+[ARTstor](http://www.artstor.org/) - More than two million images from historical, carographic, and art history sources.
```
I was surprised to see Artstor on here because I wasn't aware they had an API for accessing their library... and I actually couldn't find one on their site after paging around a bit. Is there a more specific link to an API for this? Or should it be replaced by another API service like Europeana?

walshbr replied:
Resolved.

mdlincoln reviewed (changes requested):
I switched a few full PH urls to relative PH urls. And I had one question about Artstor...

walshbr commented:
Pinging @smythp so that he can hop in and take a look at these small things. They're all pretty minor tweaks, so hopefully should be a quick review!

mdlincoln commented:
Also I'd echo @ianmilligan1 this lesson reads so well and I'm legit excited to learn how to create an API rather than just consume it 😄

smythp commented:
Hi Ian and Matthew,

Thanks for the review—I'm excited that the tutorial is almost ready to go up on the site. To the extent that it reads well, I can only credit Brandon's hard work and the excellent feedback from my reviewers.

- The spotlight change is spot on. I'm generally on Linux so some of the Mac OS terminology escapes me.
- Looks like there are six places I use OSX and not Mac OS. Makes sense to replace these.
- I think Artstor could be removed and replaced with Europeana (https://pro.europeana.eu/)
- Data-entries is actually data—entries, with an em dash, which reads OK if you realize it but which might be confusing. I'm fine with the change to "data entries" or to leave it as is.
- I generally avoid idiomatic language like *sine qua non* in technical writing but I usually allow myself a few touches which I think breathe some personality into the writing here and there. I'd personally leave *sine qua non* and the pretty colloquial "Whew!" I use elsewhere just for that reason. I'm fine with the proposed change, though, if it doesn't fit PH style.
- I like the Twitterbot link, and I've tried to link to other PH lessons elsewhere in the tutorial, so it fits well.

Brandon, sorry if I'm missing a way to check out the PR code and make changes myself but I'm guessing I don't have permission. Let me know if I can clarify anything or if there's a way for me to add to the commits in this PR. I just have to say that I love this review procedure—I'm going to be spoiled when I go into my next peer review process. Every journal should have continuous integration...


Best,
Patrick

ianmilligan1 commented:
Sounds good @smythp – and happy for you to keep in the colloquial language too. It's a fine line to walk which you're aware of so I'm personally happy to defer to you as the author!

walshbr committed:
Spotlight

walshbr committed:
Update osx to macos

walshbr committed:
Resolve quotes

walshbr committed:
Replace Artstor

walshbr committed:
Update for twitterbot link

walshbr committed:
Spacing

walshbr committed:
Add Matt and Ian as reviewers

walshbr commented:
I wasn't sure about the permissions either @smythp. I went through and implemented the suggestions you made @ianmilligan1 and @mdlincoln - also added you as reviewers since you've been making comments on here, but let me know if you'd prefer to be left off? This should be good to go if you want to give the 👍 @mdlincoln

mdlincoln commented:
@walshbr I appreciate it but I don't think I need to be named as a reviewer - my contribution isn't really akin to a peer review

walshbr committed:
Remove Matt as reviewer

walshbr commented:
10/4 - took you off. Left @ianmilligan1 on there for the moment, but I can always take him off if he feels the same way. So let's wait to hear from him before merging.

walshbr committed:
Remove ian as review

walshbr commented:
Removed! Should be good to go.
